### PR TITLE
Remove access to field slices from composite and interface types

### LINKED ISSUE #2952: Remove access to fields by index
### Issue to be solved

Accessing fields by index (instead of by name) can create dependency on sort order.

For example, issue https://github.com/onflow/cadence/issues/2558:
> Some client programs are accessing fields by index rather than name. This created a dependency on the sort order of fields which surfaced when we switched from JSON-CDC to CCF for encoding events because CCF explicitly sorts as specified in CCF specification for deterministic encodings.

PR https://github.com/onflow/cadence/pull/2559 resolved issue 2558 by updating CCF Codec to make CCF Determinisitic Encoding optional.  It was a stop-gap solution to avoid abruptly breaking compatibility for encoded events.

For exported values, not allowing access to fields by index can help prevent Cadence developers from unintentionally creating dependency on sort order.  This can prevent unintended consequences like blocking deterministic encoding, etc.

However, this issue can be viewed as broader topic than CCF, sort order, or access by field index.

### Current Status

CCF encoding is not currently sorting the fields (in order to be compatible with old programs that access fields by index).

- CCF specs ([onflow/ccf](https://github.com/onflow/ccf)) defines how to sort and makes it optional, so each CCF-based format can balance trade-offs by specifying which CCF options they use:
  > CCF defines "Deterministic CCF Encoding Requirements" and makes it optional. 

- CCF codec (onflow/cadence) offers sorting as non-default option, but sorting isn't being used (e.g. for events encoding) due to compatibility reasons mentioned in:
  - #2558
  - #2559
  
I didn't find any programs using the sort option in CCF codec (as of Nov 28, 2023).

### Suggested Solution

Programming languages before 1.0 can benefit from taking away the ability for programmers to (sometimes unintentionally) create dependencies on implementation details of compilers and interpreters.

For example, a few months before Go 1.0 (~12 years ago), https://codereview.appspot.com/5285042/ added a random offset for map iteration and changed Go language specification.  

```diff
Index: doc/go_spec.html
===================================================================
--- a/doc/go_spec.html
+++ b/doc/go_spec.html
@@ -4085,7 +4085,8 @@
 </li>
 
 <li>
-The iteration order over maps is not specified.
+The iteration order over maps is not specified
+and is not guaranteed to be the same from one iteration to the next.
 If map entries that have not yet been reached are deleted during iteration,
 the corresponding iteration values will not be produced. If map entries are
 inserted during iteration, the behavior is implementation-dependent, but the
```

That change to Go intentionally broke some existing Go programs by taking away existing behavior that wouldn't be ideal to officially or unofficially support forever after Go 1.0 release.

In the same way, maybe we can determine if some existing features (accessing field by index) should be removed before Cadence 1.0 release.

EDIT: added Current Status section about CCF for more context.

### PR DESCRIPTION
**Pull request by turbolent**
Closes #2952

## Description

Similar to how access to field value slices was from composite values was previously removed in #3290, also remove access to field slices of composite and interface types.

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: values.go [hunk 1]
```diff
@@ -1521,16 +1521,64 @@ func NewMeteredKeyValuePair(gauge common.MemoryGauge, key, value Value) KeyValue
...
+func getCompositeFieldValues(composite Composite) []Value { //nolint:unused
 	return composite.getFieldValues()
 }
 
+// SearchFieldByName searches for the field with the given name in the composite type,
+// and returns the value of the field, or nil if the field is not found.
+//
+// WARNING: This function performs a linear search, so is not efficient for accessing multiple fields.
+// Prefer using FieldsMappedByName if you need to access multiple fields.
+func SearchFieldByName(v Composite, fieldName string) Value {
```
Comment (turbolent) [inline values.go:1544]: Moved as-is from `types.go`

File: values.go [hunk 2]
```diff
@@ -1521,16 +1521,64 @@ func NewMeteredKeyValuePair(gauge common.MemoryGauge, key, value Value) KeyValue
...
+
+	for i, field := range fields {
+		if field.Identifier == fieldName {
+			return fieldValues[i]
+		}
+	}
+	return nil
+}
+
+func FieldsMappedByName(v Composite) map[string]Value {
```
Comment (turbolent) [inline values.go:1560]: Moved as-is from `types.go`

File: values.go [hunk 3]
```diff
@@ -1521,16 +1521,64 @@ func NewMeteredKeyValuePair(gauge common.MemoryGauge, key, value Value) KeyValue
...
 type Composite interface {
 	Value
-	GetFields() []Field
+
+	isComposite()
+	getFields() []Field
 	getFieldValues() []Value
+
+	SearchFieldByName(fieldName string) Value
+	FieldsMappedByName() map[string]Value
```
Comment (turbolent) [inline values.go:1530]: Feedback from users: Expose existing global functions with the same name as methods on `Composite` for convenience.

**Review by fxamacker — APPROVED** [review]
Thanks for removing access to field types!  LGTM!

With both field types and values unexported, is it safe from Cadence perspective to sort fields in CCF now?

File: values.go [hunk 4]
```diff
@@ -1521,16 +1521,64 @@ func NewMeteredKeyValuePair(gauge common.MemoryGauge, key, value Value) KeyValue
...
+
+	fieldsMap := make(map[string]Value, len(fields))
+	for i, fieldValue := range fieldValues {
+		var fieldName string
+		if i < len(fields) {
+			fieldName = fields[i].Identifier
+		} else if attachment, ok := fieldValue.(Attachment); ok {
+			fieldName = interpreter.AttachmentMemberName(attachment.Type().ID())
+		}
+		fieldsMap[fieldName] = fieldValue
```
[the remark this answers is not in the document]
Comment (turbolent) [reply values.go:1576]: I guess so? I'll add a panic just in case, good idea

**Review by SupunS — APPROVED** [review]

**Pull request merged**